src/Lykke.Service.BitfinexAdapter.Services/Exchange/BitfinexExchange.cs

```diff
@@ -0,0 +1,238 @@
+{
+    public class BitfinexExchange : ExchangeBase
+    {
+        private readonly BitfinexModelConverter _modelConverter;
+        private readonly BitfinexOrderBooksHarvester _orderBooksHarvester;
+        //private readonly BitfinexExecutionHarvester _executionHarvester;
+        private readonly IBitfinexApi _exchangeApi;
+
+        public BitfinexExchange(BitfinexAdapterSettings configuration,
+            //TranslatedSignalsRepository translatedSignalsRepository,
+            BitfinexOrderBooksHarvester orderBooksHarvester,
+            //BitfinexExecutionHarvester executionHarvester, 
+            ILog log)
+            : base(Constants.BitfinexExchangeName, configuration, log)
+        {
+            _modelConverter = new BitfinexModelConverter(configuration);
+            _orderBooksHarvester = orderBooksHarvester;
+            //_executionHarvester = executionHarvester;
+            var credenitals = new BitfinexServiceClientCredentials(configuration.ApiKey, configuration.ApiSecret);
+            _exchangeApi = new BitfinexApi(credenitals)
+            {
+                BaseUri = new Uri(configuration.EndpointUrl)
+            };
+
+
+
+            orderBooksHarvester.MaxOrderBookRate = configuration.MaxOrderBookRate;
+        }
+
+        public override async Task<ExecutionReport> AddOrderAndWaitExecution(TradingSignal signal, TimeSpan timeout)
+        {
+            var symbol = _modelConverter.LykkeSymbolToExchangeSymbol(signal.Instrument.Name);
+            var volume = signal.Volume;
+            var orderType = _modelConverter.ConvertOrderType(signal.OrderType);
+            var side = _modelConverter.ConvertTradeType(signal.TradeType);
+            var price = signal.Price == 0 ? 1 : signal.Price ?? 1;
+            var cts = new CancellationTokenSource(timeout);
+
+            var response = await _exchangeApi.AddOrder(symbol, volume, price, side, orderType, cts.Token);
+
+            if (response is Error error)
+            {
+                throw new ApiException(error.Message);
+            }
+
+            var trade = OrderToTrade((Order)response);
+            return trade;
+        }
+
+        public override async Task<ExecutionReport> CancelOrderAndWaitExecution(TradingSignal signal, TimeSpan timeout)
+        {
+
+            var cts = new CancellationTokenSource(timeout);
+            if (!long.TryParse(signal.OrderId, out var id))
+            {
+                throw new ApiException("Bitfinex order id can be only integer");
+            }
+            var response = await _exchangeApi.CancelOrder(id, cts.Token);
+            if (response is Error error)
+            {
+                throw new ApiException(error.Message);
+            }
+            var trade = OrderToTrade((Order)response);
+            return trade;
+        }
+
+        public override async Task<ExecutionReport> GetOrder(string id, Instrument instrument, TimeSpan timeout)
+        {
+            if (!long.TryParse(id, out var orderId))
```

Comment:
Looks like copy-paste, extract to some method

Reply:
This has been refactored and will be reviewed in the next PR (its in a different branch)